src/lib/group.ts

```diff
@@ -86,32 +116,102 @@ class Group {
   add(g: Group) {
     if (this.#isConstant() && g.#isConstant()) {
-      if (this.x.toBigInt() === 0n) {
+      // we check if either operand is zero, because adding zero to g just results in g (and vise versa)
+      if (this.#isZero().toBoolean()) {
         return g;
-      } else if (g.x.toBigInt() === 0n) {
+      } else if (g.#isZero().toBoolean()) {
         return this;
       } else {
         let g_proj = Pallas.add(this.#toProjective(), g.#toProjective());
-        return Group.#fromProjective(g_proj);
+
+        // in the JS code, zero is denoted with (1, 1) - but here we want to convert it to (0, 0) (it saves constraints in the in-circuit implementation that way)
+        let isZero = g_proj.x === 1n && g_proj.y === 1n;
+        return isZero
+          ? new Group({ x: 0, y: 0 })
+          : Group.#fromProjective(g_proj);
       }
     } else {
-      let [, x, y] = Snarky.group.add(this.#toTuple(), g.#toTuple());
-      return new Group({ x, y });
+      const { x: x1, y: y1 } = this;
+      const { x: x2, y: y2 } = g;
+
+      let zero = new Field(0);
+
+      let same_x = Provable.witness(Field, () => x1.equals(x2).toField());
+
+      let inf = checkFinite
+        ? zero
+        : Provable.witness(Field, () =>
+            x1.equals(x2).and(y1.equals(y2).not()).toField()
+          );
+
+      let inf_z = Provable.witness(Field, () => {
+        if (y1.equals(y2).toBoolean()) return zero;
+        else if (x1.equals(x2).toBoolean()) return y2.sub(y1).inv();
+        else return zero;
+      });
+
+      let x21_inv = Provable.witness(Field, () => {
+        if (x1.equals(x2).toBoolean()) return zero;
+        else return x2.sub(x1).inv();
+      });
+
+      let s = Provable.witness(Field, () => {
+        if (x1.equals(x2).toBoolean()) {
+          let x1_squared = x1.square();
+          return x1_squared.add(x1_squared).add(x1_squared).div(y1.add(y1));
+        } else return y2.sub(y1).div(x2.sub(x1));
+      });
+
+      let x3 = Provable.witness(Field, () => {
+        return s.square().sub(x1.add(x2));
+      });
+
+      let y3 = Provable.witness(Field, () => {
+        return s.mul(x1.sub(x3)).sub(y1);
+      });
+
+      let [, x, y] = Snarky.group.ecadd(
+        Group.from(x1.seal(), y1.seal()).#toTuple(),
+        Group.from(x2.seal(), y2.seal()).#toTuple(),
+        Group.from(x3, y3).#toTuple(),
+        inf.value,
+        same_x.value,
+        s.value,
+        inf_z.value,
+        x21_inv.value
+      );
+
+      // similarly to the constant implementation, we check if either operand is zero
+      // and the implementation above (original OCaml implementation) returns something wild -> g + 0 != g where it should be g + 0 = g
+      let gIsZero = g.#isZero();
+      let thisIsZero = this.#isZero();
+
+      // if either one is the negation of the other, we just return the zero element since g + (-g) = 0 - but the OCaml implementation doesn't pick that up
+      let isNegation = g.neg().equals(this);
```

Comment:
this should just be set to `inf` if it's not hard-coded anymore

afaict, `inf` will be 0, and constrained to 0, if either `this` of `g` is 0